Fix down migration for build event sequences

clarafu opened the pull request:
## Changes proposed by this PR

#7857

There was a typo in the down migration for constructing build event sequences in postgres but it was never caught because the migration only runs for started builds during the downgrade but the integration tests don't cover that use case.

* [x] done
* [ ] todo

## Release Note

* This PR fixes a typo in a down migration that affects Concourse version 7.4.1, 7.4.2 and 7.4.3. Therefore, you will be unable to downgrade from any of those listed versions.

clarafu committed:
atc: fix down migration

There was a typo in the down migration. It never got properly tested
because the downgrade integration tests never actually runs these down
migrations since they are only run if there are started builds during
the downgrade.

[#7857]

Signed-off-by: Clara Fu <fclara@vmware.com>

Berber31 commented:
what do you think about adding some tests to cover thoses lines ?
as mentioned here: https://github.com/concourse/concourse/issues/7857#issuecomment-985606916

Berber31 commented:
@clarafu 
i think the query is too restrictive

`select id, name, pipeline_id, team_id from builds where status = 'started' and completed = false and aborted = false`

i suggest to remove `aborted = false` or add `or end_time is null`

we could face off a build aborted after a migration to 7.4.1. So aborted column could be 'true', so sequence will not be recreated.

we have this situation actually

 status  | completed | aborted | end_time |    id    | build_number |                  job_name                  |             name             |   name   
---------+-----------+---------+----------+----------+--------------+--------------------------------------------+------------------------------+----------
 started | f         | t       |          | 70973927 | 17           | controle_validite_changement_en_production | mars-acquisitionrecouvrement | der-grea


with theses logs:
```

concourse-postgresql 2021-12-07 11:15:48.696 GMT [4011] ERROR:  relation "build_event_id_seq_70973927" does not exist at character 96                                                                           │
│ concourse-postgresql 2021-12-07 11:15:48.696 GMT [4011] STATEMENT:  INSERT INTO pipeline_build_events_337 (event_id,build_id,type,version,payload) VALUES (nextval('build_event_id_seq_70973927'),$1,$2,$3,$4)
```

Berber31 commented:
> ecause this is such a specific use case that adding one for it in the



> @Berber31 I think the reasoning behind that is because if the build is aborted then we won't need to apply more events to that build, therefore we don't need to migrate that event sequence?

So if a build started on 7.4.0 then aborted in 7.4.2 (and migrate DB back to 7.4.0) we won't be able to abort this build and will run forever (with the yellow highlight still blinking on this job)

clarafu commented:
Hm I see, you mean the build is set as aborted but the build itself has not actually been successfully been aborted yet and still continues to have some output. I see your case then, I agree I will change the query. Thanks!!

clarafu committed:
db: modify down migration to loosen requirements

Berber31 brought up a good point that there can be builds that have been
aborted but not yet completed that still require some build output.
Therefore the old down migration query would not have created build
event sequences for those builds and those "aborted" builds would error
because it does not have event sequences after downgrading. This commit
loosens up the restrictions by creating event sequences for even aborted
builds as long as the build is not completed yet

[#7857]

Signed-off-by: Clara Fu <fclara@vmware.com>

clarafu committed:
db: further improve down migration

The down migration also does not over pending builds. Previously, we
would create build event sequences once we call createBuild and these
builds start off with a pending status. This means it is possible for
there to be pending builds that do not have the expected build event
sequence when we downgrade, which this commit will fix by constructing
event sequences for both started and pending builds.

[#7857]

Signed-off-by: Clara Fu <fclara@vmware.com>

clarafu commented:
@xtremerui Could you review it one more time! Just for a sanity check haha because I made some changes

xtremerui commented:
@clarafu just checked again I think it is fine to me.

jmccann commented:
Was wondering if anyone could post some notes to https://github.com/concourse/concourse/discussions/8298 regarding how to upgrade from 7.4.3 to 7.7.1 with these DB migrations in mind.
